### PR TITLE
Win checkers

### PR DESCRIPTION
**Pull request by Haremun**
User can win and draw now. There are all win checkers. Functions are rather long but I will fix it in next commits.

### CODE REVIEW

File: src/main/java/com/bieganski/ox/GameFlow.java [hunk 1]
```diff
@@ -31,14 +31,14 @@ void run() {
     //TODO start player should be from game setup
     currentSymbol = Symbol.values()[0];
 
-    makeTurn(currentSymbol);
+    makeTurn();
```
Comment (hongshui) [inline src/main/java/com/bieganski/ox/GameFlow.java]: You do not 'make a turn'

File: src/main/java/com/bieganski/ox/GameFlow.java [hunk 2]
```diff
@@ -31,14 +31,14 @@ void run() {
...
   }
 
-  private void makeTurn(Symbol currentSymbol) {
+  private void makeTurn() {
```
Comment (hongshui) [inline src/main/java/com/bieganski/ox/GameFlow.java]: Less parameters, nice to see :)

File: src/main/java/com/bieganski/ox/checkers/HorizontalWinChecker.java [hunk 3]
```diff
@@ -4,45 +4,42 @@
...
+      if (upper) {
+        field = correct.last();
+        nextField = helper.getHigherField(field, field.hashCode() + 1);
+        upper = appendFieldIfCorrect(field, nextField, correct);
+      }
+      if (lower) {
+        field = correct.first();
+        nextField = helper.getLowerField(field, field.hashCode() - 1);
+        lower = appendFieldIfCorrect(field, nextField, correct);
+      }
```
Comment (hongshui) [inline src/main/java/com/bieganski/ox/checkers/HorizontalWinChecker.java:36]: This two are very similar, could be extracted to separate method, You can use functional interfaces to do this

File: src/main/java/com/bieganski/ox/checkers/HorizontalWinChecker.java [hunk 4]
```diff
@@ -4,45 +4,42 @@
...
-
-  private boolean checkFieldsAreNotNull(Field current, Field next) {
-    return current != null && next != null;
-  }
-
-  private boolean isDistanceCorrectAndSymbolsAreEqual(Field current, Field next) {
-    return current.distanceTo(next) <= 1 && current.areSymbolsEqual(next);
+  protected boolean isDistanceCorrectAndSymbolsAreEqual(Field current, Field next) {
+    return current.areSymbolsEqual(next)
+        && current.hashCode() / boardSize == next.hashCode() / boardSize;
```
Comment (hongshui) [inline src/main/java/com/bieganski/ox/checkers/HorizontalWinChecker.java]: This one is rather confusing. Consider changing it

File: src/main/java/com/bieganski/ox/checkers/Judge.java [hunk 5]
```diff
@@ -34,9 +37,13 @@ public void onFieldAdded(TreeSet<Field> fieldsWithValue, Field addedField, int s
...
     }
+    if (fieldsWithValue.size() == size * size) {
+      App.LOG.info("Draw! All fields all occupied");
+      gameListener.onDraw();
+    }
   }
 
   private boolean checkWin(TreeSet<Field> fieldsWithValue, Field addedField, int size) {
-    return checkers.stream().anyMatch(x -> x.checkWin(fieldsWithValue, addedField, boardSize));
+    return checkers.stream().anyMatch(x -> x.checkWin(fieldsWithValue, addedField));
```
Comment (hongshui) [inline src/main/java/com/bieganski/ox/checkers/Judge.java:47]: Nice one :)

File: src/main/java/com/bieganski/ox/checkers/ReverseSlantWinChecker.java [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+        upper = appendFieldIfCorrect(field, nextField, correct);
+      }
+      if (lower) {
+        field = correct.first();
+        nextField = helper.getLowerField(field, field.hashCode() - boardSize - 1);
+        lower = appendFieldIfCorrect(field, nextField, correct);
+      }
+    }
+    return correct.size() >= winSize;
+  }
```
Comment (hongshui) [inline src/main/java/com/bieganski/ox/checkers/ReverseSlantWinChecker.java:36]: Duplication of code from previous classes

File: src/main/java/com/bieganski/ox/checkers/ReverseSlantWinChecker.java [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+
+  protected boolean isDistanceCorrectAndSymbolsAreEqual(Field current, Field next) {
+    return current.areSymbolsEqual(next)
+        && (current.hashCode() / boardSize) != (next.hashCode() / boardSize);
+  }
```
Comment (hongshui) [inline src/main/java/com/bieganski/ox/checkers/ReverseSlantWinChecker.java]: Do not use 'and' in method names, this screams "breaking SRP"

File: src/main/java/com/bieganski/ox/checkers/SlantWinChecker.java [hunk 8]
```diff
@@ -0,0 +1,42 @@
...
+        upper = appendFieldIfCorrect(field, nextField, correct);
+      }
+      if (lower) {
+        field = correct.first();
+        nextField = helper.getLowerField(field, field.hashCode() - boardSize - 1);
+        lower = appendFieldIfCorrect(field, nextField, correct);
+      }
+    }
+    return correct.size() >= winSize;
+  }
```
Comment (hongshui) [inline src/main/java/com/bieganski/ox/checkers/SlantWinChecker.java:36]: Again, long, complicated and duplicated

File: src/main/java/com/bieganski/ox/checkers/TreeSetHelper.java [hunk 9]
```diff
@@ -0,0 +1,40 @@
...
+    if (next != null) {
+      if (next.hashCode() == position) {
+        return next;
+      }
+      if (next.hashCode() > position) {
+        return getLowerField(next, position);
+      }
+    }
+    return null;
+  }
```
Comment (hongshui) [inline src/main/java/com/bieganski/ox/checkers/TreeSetHelper.java:39]: Duplicated

**Review by hongshui — APPROVED** [review]
A lot of code duplication. Try to change it in the future. But for now it can stay like this.

**Comment by Haremun** [comment]
Just changed some things from @hongshui review.

**Pull request merged**